File: ch15/Chapter15-CustomProviders-VB/App_Code/LimitedSqlRoleProvider.vb
Imports System.Web.Security
Imports System.Configuration.Provider

Public Class LimitedSqlRoleProvider
    Inherits SqlRoleProvider

    Public Overrides Sub CreateRole(ByVal roleName As String)
        If (roleName = "Administrator" Or roleName = "Manager") Then
            MyBase.CreateRole(roleName)
        Else
            Throw New  _
             ProviderException("Role creation limited to only Administrator and Manager")
        End If
    End Sub

    Public Overrides Function DeleteRole(ByVal roleName As String, _
        ByVal throwOnPopulatedRole As Boolean) As Boolean
        Return False
    End Function

    Public Overrides Sub AddUsersToRoles(ByVal usernames() As String, _
        ByVal roleNames() As String)

        For Each roleItem As String In roleNames
            If roleItem = "Administrator" Then
                Throw New  _
                    ProviderException("You are not authorized to add any users" & _
                        " to the Administrator role")
            End If
        Next

        MyBase.AddUsersToRoles(usernames, roleNames)
    End Sub

End Class

File: ch15/Chapter15-CustomProviders-CS/App_Code/LimitedSqlRoleProvider.cs
using System;
using System.Collections.Generic;
using System.Configuration.Provider;
using System.Linq;
using System.Web;
using System.Web.Security;

namespace Chapter15_CustomProviders_CS.App_Code
{
    public class LimitedSqlRoleProvider : SqlRoleProvider
    {
        public override void CreateRole(string roleName)
        {
            if (roleName == "Administrator" || roleName == "Manager")
            {
                base.CreateRole(roleName);
            }
            else
            {
                throw new
                    ProviderException("Role creation limited to only Administrator and Manager");
            }
        }

        public override bool DeleteRole(string roleName, bool throwOnPopulatedRole)
        {
            return false;
        }

        public override void AddUsersToRoles(string[] usernames, string[] roleNames)
        {
            foreach (string roleItem in roleNames)
            {
                if (roleItem == "Administrator")
                {
                    throw new ProviderException("You are not authorized to add any users" +
                        " to the Administrator role");
                }
            }

            base.AddUsersToRoles(usernames, roleNames);
        }
    }
}